### ISSUE TITLE
YOLO26n导出tflite后，索引为0的类别会出现与上一条类别定位完全重合的错误

### ISSUE THREAD

**Issue by Starkguava** [ultralytics/ultralytics#23346]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

YOLO26n导出tflite后，索引为0的类别会出现与上一条类别定位完全重合的错误

### Additional

我们简单训练了一个ocr识别数字和字母的demo，在.pt下测试是完全正确的，而在导出tflite后会出现以下问题：
0 = {float[6]@47206} [0.104787946, 0.46824303, 0.18168591, 0.58257985, 0.96063465, 24.0]
**1 = {float[6]@47207} [0.019556295, 0.4436751, 0.107705675, 0.5854393, 0.9603612, 23.0]
2 = {float[6]@47208} [0.019556295, 0.4436751, 0.107705675, 0.5854393, 0.9579195, 0.0]
3 = {float[6]@47209} [0.28735766, 0.4436547, 0.36102203, 0.56042284, 0.95364594, 3.0]
4 = {float[6]@47210} [0.28735766, 0.4436547, 0.36102203, 0.56042284, 0.9529765, 0.0]**
5 = {float[6]@47211} [0.38027313, 0.44484878, 0.4585964, 0.56000495, 0.95124876, 0.0]
6 = {float[6]@47212} [0.47619057, 0.44451082, 0.5499851, 0.5611348, 0.95100635, 8.0]
7 = {float[6]@47213} [0.8529782, 0.44925645, 0.93138236, 0.5644056, 0.9507863, 7.0]
8 = {float[6]@47214} [0.6650287, 0.44675952, 0.7398499, 0.5633069, 0.94486284, 7.0]
9 = {float[6]@47215} [0.1915837, 0.44322687, 0.2707733, 0.5611132, 0.9418906, 2.0]
后面的省略

<img width="721" height="721" alt="Image" src="https://github.com/user-attachments/assets/742bd312-dd52-41fa-8e53-58f731c70903" />

以下是我处理后的结果：
0 = {Recognition@47194} Recognition(title=O, confidence=0.96063465, location=Rect(76, 338 - 131, 420))
1 = {Recognition@47195} Recognition(title=N, confidence=0.9603612, location=**Rect(14, 320 - 78, 422)**)
2 = {Recognition@47196} Recognition(title=0, confidence=0.9579195, location=**Rect(14, 320 - 78, 422)**)
3 = {Recognition@47197} Recognition(title=3, confidence=0.95364594, location=**Rect(207, 320 - 260, 404)**)
4 = {Recognition@47198} Recognition(title=0, confidence=0.9529765, location=**Rect(207, 320 - 260, 404)**)
5 = {Recognition@47199} Recognition(title=0, confidence=0.95124876, location=Rect(274, 321 - 331, 404))
6 = {Recognition@47200} Recognition(title=8, confidence=0.95100635, location=Rect(343, 320 - 397, 405))
7 = {Recognition@47201} Recognition(title=7, confidence=0.9507863, location=Rect(615, 324 - 672, 407))
8 = {Recognition@47202} Recognition(title=7, confidence=0.94486284, location=Rect(479, 322 - 533, 406))
9 = {Recognition@47203} Recognition(title=2, confidence=0.9418906, location=Rect(138, 320 - 195, 405))

<img width="721" height="721" alt="Image" src="https://github.com/user-attachments/assets/dd9ca592-2da8-42c3-8ff0-7d4589a5e6fc" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Starkguava, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This looks like a potential 🐛 export/inference mismatch after converting to TFLite. To help our team investigate, please share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) that reproduces the “class index 0 boxes overlap with previous class” behavior, including the exact export + TFLite inference steps and a small sample input image.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

Also note: this is an automated response 🤖, and an Ultralytics engineer will also assist soon.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
What's the output from Ultralytics inference?

**Comment by Starkguava** [comment]
您指的是.pt文件在python使用model.predict的输出么？他的输出是正确的，
![Image](https://github.com/user-attachments/assets/512f338f-b0c2-4c65-964a-ad5b271daf0b)

tflite的输出是tensor: float32[1,300,6]

**Comment by Y-T-G** [comment]
I think it's a bug in TFLite then. What was the export command?

**Comment by Starkguava** [comment]
model.export(format="tflite", half=True, device="cpu")

**Comment by Starkguava** [comment]
是否和以下warming有关？
ONNX: starting export with onnx 1.19.1 opset 22...
.venv/lib/python3.12/site-packages/torch/onnx/_internal/torchscript_exporter/utils.py:1447: OnnxExporterWarning: Exporting to ONNX opset version 22 is not supported. by 'torch.onnx.export()'. The highest opset version supported is 20. To use a newer opset version, consider 'torch.onnx.export(..., dynamo=True)'. 
 warnings.warn(
/.venv/lib/python3.12/site-packages/torch/onnx/_internal/torchscript_exporter/symbolic_opset9.py:5353: UserWarning: Exporting aten::index operator of advanced indexing in opset 22 is achieved by combination of multiple ONNX operators, including Reshape, Transpose, Concat, and Gather. If indices include negative values, the exported graph will produce incorrect results.
 warnings.warn(
ONNX: slimming with onnxslim 0.1.74...
ONNX: export success ✅ 0.6s, saved as 'model.onnx' (9.6 MB)
TensorFlow SavedModel: starting TFLite export with onnx2tf 1.28.3...
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1768965132.244704   83247 [devices.cc:76](http://devices.cc:76/)] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0 (Note: TensorFlow was not compiled with CUDA or ROCm support)
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1768965132.245285   83247 single_[machine.cc:374](http://machine.cc:374/)] Starting new session
Saved artifact at 'model_saved_model'. The following endpoints are available:

* Endpoint 'serving_default'

**Comment by Y-T-G** [comment]
You can try exporting with this:

```
pip install git+https://github.com/ultralytics/ultralytics@topk-simplify
```

**Comment by Y-T-G** [comment]
Actually, this is due to multi-label prediction. So the same box has multiple classes that are above the confidence threshold.

**Comment by Starkguava** [comment]
> 你可以尝试用这个方法导出：
> 
> ```
> pip install git+https://github.com/ultralytics/ultralytics@topk-simplify
> ```

用这个分支导出的模型是正确的了

**Comment by Y-T-G** [comment]
The branch disables multi-label prediction and only selects the best class.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in the linked PR: [Support `max_det` and `agnostic_nms` for end2end (ultralytics 8.4.8)](https://github.com/ultralytics/ultralytics/pull/23396).

### What changed (and why it may fix this)
This PR improves end-to-end (NMS-free) inference behavior by ensuring `max_det` and `agnostic_nms` are actually applied when `end2end=True`, rather than being ignored in parts of the pipeline. Key updates include:

- Added a safe, generic `set_head_attr(**kwargs)` helper to set head (last-layer) attributes like `end2end`, `max_det`, and `agnostic_nms`.
- When `end2end=True`, both **predict** and **val** now push `max_det` and `agnostic_nms` into the model head so the settings take effect as expected.
- Added `agnostic_nms` support to the `Detect` head (affecting the top-k selection logic used in end-to-end flows).
- Export pipeline now passes `agnostic_nms` into the exported head to better match Python inference behavior after export.
- Training/attribute setup sets `max_det` for end-to-end models.
- Docs/tests updated accordingly, and the package version is bumped **8.4.7 → 8.4.8**.

This should improve consistency across **predict**, **validate**, **train-time validation**, and **exported** deployments for end-to-end models (including YOLO26 and YOLOv10).

### Credits
Thanks to @Y-T-G for the PR, with contributions from @Laughing-q and @glenn-jocher.

### How to test the fix
- Upgrade to the version mentioned in the PR:
  - `pip install -U ultralytics>=8.4.8`
- Or test the latest main branch directly:
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`
- Alternatively, waiting for the next official PyPI release is an option if `8.4.8` is not yet available in the environment being used.

### Feedback requested
Any confirmation on whether these changes resolve the issue on the affected setup would be really helpful. If the issue persists, sharing a minimal repro (model, args used like `end2end/max_det/agnostic_nms`, and the observed vs expected behavior) will help narrow it down further.

Thanks 🙏 for reporting this—further feedback is welcome if anything still looks off.

**Comment by Y-T-G** [comment]
You can export with `agnostic_nms=False` with latest Ultralytics to avoid this issue.

**Comment by HonestyBrave** [comment]
> You can export with `agnostic_nms=False` with latest Ultralytics to avoid this issue.

If agnostic_nms=False is set, non-maximum suppression is only applied per class—this means overlapping bounding boxes of different classes won’t be filtered out.
Given your requirement to have only one detection box per digit, the correct setting should actually be agnostic_nms=True. or i understand not right?

**Comment by Y-T-G** [comment]
Yes, you're correct. My bad.

**Comment by Starkguava** [comment]
我在yolo26obb中好像又遇到了相同的问题 （最新版本的ultralytics包）

**Comment by Y-T-G** [comment]
Does `agnostic_nms=False` not work anymore?

**Comment by Starkguava** [comment]
agnostic_nms=True work

**Comment by glenn-jocher** [comment]
Thanks for confirming — if `agnostic_nms=True` fixes it for `yolo26obb`, that suggests the same overlapping multi-class selection behavior in the exported end-to-end output; please use that as the current workaround, and if you can still reproduce it on the latest package, open a new `OBB` issue with a minimal repro and `yolo checks` output so we can track it separately, and see the [end-to-end detection guide](https://docs.ultralytics.com/guides/end2end-detection/) for the expected export behavior.
